### DISCUSSION TITLE
error report in the constructor

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#22709] [Q&A General] [answered]
Hi expert,

could you please help to see how to solve the following problems.
I have 3 velocity component but in the construtor it generates an error.
Perhaps I forgot to include some h file? Thank you so much.

I have following in my constructor:
  _u_x_l(adCoupledValue("u_x_l")),
  _u_y_l(_mesh.dimension() >= 2.0 ? adCoupledValue("u_y_l") : _zero),

The reported error is 
_error: operands to '?:' have different types 'const ADVariableValue' {aka 'const MooseArray<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> >'} and 'const VariableValue' {aka 'const MooseArray<double>'}
_u_y_l(_mesh.dimension() >= 2.0 ? adCoupledValue("u_y_l") : _zero),

Best regards,
Qiao

**Comment by GiudGiud** [comment, marked as answer]
Hello

_zero isnt the right value for ad variable values

you want to use _ad_zero

Guillaume

**Reply by Oops-Qiao** [reply]
> Hello
> 
> _zero isnt the right value for ad variable values
> 
> you want to use _ad_zero
> 
> Guillaume

Thanks very much my dear friend!!!

**Reply by Oops-Qiao** [reply]
> Hello
> 
> _zero isnt the right value for ad variable values
> 
> you want to use _ad_zero
> 
> Guillaume

Hello Guillaume, I have a quick question,
in the AD version I saw two different ways to define the parameters, what are their differences? Thank you.

1)  Real _ambient_convection_alpha; 
https://github.com/idaholab/moose/blob/next/modules/navier_stokes/include/materials/INSAD3Eqn.h

 2)   ADReal _hmax;
https://github.com/idaholab/moose/blob/next/modules/navier_stokes/include/materials/INSADTauMaterial.h


Qiao

**Reply by GiudGiud** [reply]
The difference is whether the number carries derivative information or not.

A constant should pretty much always be a Real not an ADReal, though it s often inconsequential to have it be an ADReal

**Reply by Oops-Qiao** [reply]
ok, thanks very much.
